crates/client/mempool/src/inner/limits.rs

```diff
@@ -0,0 +1,148 @@
+    pub fn new(chain_config: &ChainConfig) -> Self {
+        Self {
+            max_transactions: chain_config.mempool_tx_limit,
+            max_declare_transactions: chain_config.mempool_declare_tx_limit,
+            max_age: chain_config.mempool_tx_max_age,
+        }
+    }
+    #[cfg(any(test, feature = "testing"))]
+    pub fn for_testing() -> Self {
+        Self {
+            max_age: Duration::from_secs(10000000),
+            max_declare_transactions: usize::MAX,
+            max_transactions: usize::MAX,
+        }
+    }
+}
+
+/// Note: when a transaction is poped from the mempool by block prod, the limits will not be updated until the full
+/// tick has been executed and excess transactions are added back into the mempool.
+/// This means that the inner mempool may have fewer transactions than what the limits says at a given time.
+#[derive(Debug)]
+pub(crate) struct MempoolLimiter {
+    pub config: MempoolLimits,
+    current_transactions: usize,
+    current_declare_transactions: usize,
+}
+
+#[derive(thiserror::Error, Debug, PartialEq, Eq)]
+pub enum MempoolLimitReached {
+    #[error("The mempool has reached the limit of {max} transactions")]
+    MaxTransactions { max: usize },
+    #[error("The mempool has reached the limit of {max} declare transactions")]
+    MaxDeclareTransactions { max: usize },
+    #[error("The transaction age is greater than the limit of {max:?}")]
+    Age { max: Duration },
+}
+
+pub(crate) struct TransactionCheckedLimits {
+    check_tx_limit: bool,
+    check_declare_limit: bool,
+    check_age: bool,
+    tx_arrived_at: SystemTime,
+}
+
+impl TransactionCheckedLimits {
+    // Returns which limits apply for this transaction.
+    // This struct is also used to update the limits after insretion, without having to keep a clone of the transaction around.
+    // We can add more limits here as needed :)
+    pub fn limits_for(tx: &MempoolTransaction) -> Self {
+        match tx.tx.tx_type() {
+            TransactionType::Declare => TransactionCheckedLimits {
+                check_tx_limit: true,
+                check_declare_limit: true,
+                check_age: true,
+                tx_arrived_at: tx.arrived_at,
+            },
+            TransactionType::DeployAccount => TransactionCheckedLimits {
+                check_tx_limit: true,
+                check_declare_limit: false,
+                check_age: true,
+                tx_arrived_at: tx.arrived_at,
+            },
+            TransactionType::InvokeFunction => TransactionCheckedLimits {
+                check_tx_limit: true,
+                check_declare_limit: false,
+                check_age: true,
+                tx_arrived_at: tx.arrived_at,
+            },
+            TransactionType::L1Handler => TransactionCheckedLimits {
+                check_tx_limit: true,
+                check_declare_limit: false,
+                check_age: false,
+                tx_arrived_at: tx.arrived_at,
+            },
```

Comment:
it's because we don't want to miss l1 handler transactions - these are transactions added on the L1 core contract.

I fixed the check_tx_limit which did count for L1Handler, i missed that before. Now it still counts but won't reject any transactions for this reason